docs/src/rust-client/oracle_tutorial.md

```diff
@@ -0,0 +1,360 @@
+# Consuming On-Chain Price Data from the Pragma Oracle
+
+_Using the Pragma Oracle to get on chain price data_
+
+## Overview
+
+In this tutorial, we will build a simple “price reader” smart contract that will read Bitcoin price data from the on-chain Pragma oracle.
+
+We will use a script to call the `read_price` function in our "price reader" smart contract, which, in turn, calls the Pragma oracle via foreign procedure invocation (FPI). This tutorial lays the foundation for how you can integrate on-chain price data into your DeFi applications on Miden.
+
+## What we'll cover
+
+- Deploying a smart contract that can read oracle price data
+- Using foreign procedure invocation to get real time on-chain price data
+
+## Prerequisites
+
+This tutorial assumes you have a basic understanding of Miden assembly, have completed the previous tutorials on using the Rust client, and have completed the tutorial on foreign procedure invocation.
+
+To quickly get up to speed with Miden assembly (MASM), please play around with running Miden programs in the [Miden playground](https://0xpolygonmiden.github.io/examples/).
+
+## Step 1: Initialize your repository
+
+Create a new Rust repository for your Miden project and navigate to it with the following command:
+
+```bash
+cargo new miden-defi-app
+cd miden-defi-app
+```
+
+Add the following dependencies to your `Cargo.toml` file:
+
+```toml
+miden-client = { version = "0.8.1", features = ["testing", "concurrent", "tonic", "sqlite"] }
+miden-lib = { version = "0.8", default-features = false }
+miden-objects = { version = "0.8", default-features = false }
+miden-crypto = { version = "0.14.0", features = ["executable"] }
+miden-assembly = "0.13.0"
+rand = { version = "0.9" }
+serde = { version = "1", features = ["derive"] }
+serde_json = { version = "1.0", features = ["raw_value"] }
+tokio = { version = "1.40", features = ["rt-multi-thread", "net", "macros"] }
+rand_chacha = "0.9.0"
+miden-client-tools = "0.1.0"
+```
+
+### Step 1: Set up your `src/main.rs` file
+
+Copy and paste the following code into your `src/main.rs` file:
+
+```rust
+use rand::RngCore;
+use std::{fs, path::Path};
+
+use miden_client::{
+    account::{AccountBuilder, AccountId, AccountStorageMode, AccountType, StorageSlot, component::AccountComponent},
+    rpc::{
+        domain::account::{AccountStorageRequirements, StorageMapKey},
+        Endpoint,
+    },
+    transaction::{
+        ForeignAccount, TransactionKernel, TransactionRequestBuilder, TransactionScript,
+    },
+    Client, ClientError, Felt, Word, ZERO,
+};
+
+use miden_client_tools::{
+    instantiate_client, create_library, delete_keystore_and_store,
+};
+
+/// Import the oracle + its publishers and return the ForeignAccount list
```

Comment:
Can you spend 2 more sentences inline in the code here why we need the list of foreign accounts?